<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-БД-и-выгрузка-логов-в-файл-.csv" data-toc-modified-id="Подключение-к-БД-и-выгрузка-логов-в-файл-.csv-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к БД и выгрузка логов в файл .csv</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Вопросы-тестового-задания." data-toc-modified-id="Вопросы-тестового-задания.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вопросы тестового задания.</a></span><ul class="toc-item"><li><span><a href="#Между-какими-событиями-наибольший-шаг?-Укажите-пару-идентификаторов-событий-с-наибольшим-шагом,-относительно-всей-базы-(всех-файлов)." data-toc-modified-id="Между-какими-событиями-наибольший-шаг?-Укажите-пару-идентификаторов-событий-с-наибольшим-шагом,-относительно-всей-базы-(всех-файлов).-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Между какими событиями наибольший шаг? Укажите пару идентификаторов событий с наибольшим шагом, относительно всей базы (всех файлов).</a></span></li><li><span><a href="#Интервал-времени-до-наступления-какого-события-показывает-наибольший-разброс.-Укажите-пару-идентификатор-и-sender-с-наибольшим-разбросом.-Предложите-свои-идей,-с-чем-это-может-быть-связано." data-toc-modified-id="Интервал-времени-до-наступления-какого-события-показывает-наибольший-разброс.-Укажите-пару-идентификатор-и-sender-с-наибольшим-разбросом.-Предложите-свои-идей,-с-чем-это-может-быть-связано.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Интервал времени до наступления какого события показывает наибольший разброс. Укажите пару идентификатор и sender с наибольшим разбросом. Предложите свои идей, с чем это может быть связано.</a></span></li><li><span><a href="#Какая-задача-прогнозирования-выполняется-дольше-всего-(название-задачи)?" data-toc-modified-id="Какая-задача-прогнозирования-выполняется-дольше-всего-(название-задачи)?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Какая задача прогнозирования выполняется дольше всего (название задачи)?</a></span></li><li><span><a href="#Какое-количество-задач-прогнозирования-могут-выполняться-одновременно?-Укажите-максимальное-число-параллельных-задач." data-toc-modified-id="Какое-количество-задач-прогнозирования-могут-выполняться-одновременно?-Укажите-максимальное-число-параллельных-задач.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Какое количество задач прогнозирования могут выполняться одновременно? Укажите максимальное число параллельных задач.</a></span></li><li><span><a href="#Перечислите-forecastMarker-которые-не-завершились-успешно." data-toc-modified-id="Перечислите-forecastMarker-которые-не-завершились-успешно.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Перечислите forecastMarker которые не завершились успешно.</a></span></li></ul></li><li><span><a href="#Итоги-исследования." data-toc-modified-id="Итоги-исследования.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Итоги исследования.</a></span></li></ul></div>

In [1]:
import sqlite3 as sl
import csv
import os
import pandas as pd 
from tqdm import tqdm 

## Подключение к БД и выгрузка логов в файл .csv

In [2]:
# создадим подключение к БД с данными за 12 сентября
con = sl.connect('logs_2022-09-12.db')

In [3]:
# считаем данные из таблицы event_name
with con:
    event_name = con.execute("SELECT * FROM event_name")
    for row in event_name:
        print(row)

(1, 'Инициализируем свойства класса')
(2, 'Переход в функцию getParametrsAndRegressors')
(3, 'Выход из функции getParametrsAndRegressors с параметрами')
(4, 'Выход из функции getParametrsAndRegressors БЕЗ параметров')
(5, 'Записали в свойства класса гиперпараметры и регрессоры')
(6, 'Начинаем читать данные из BQ по полученному SQL запросу')
(7, 'Получили данные из BQ по полученному SQL запросу')
(8, 'saveForecastRequest: вход в функцию')
(9, 'saveForecastRequest: выход из функции')
(10, 'getDOW: вход в функцию')
(11, 'getDOW: выход из функции')
(12, 'getCutoffs: вход в функцию')
(13, 'getCutoffs: выход из функции')
(14, 'get_forecast: вход в функцию')
(15, 'get_forecast: выход из функции')
(16, 'updateForecastAnswer: вход в функцию')
(17, 'updateForecastAnswer: выход из функции')
(18, 'Инициализируем модель Prophet')
(19, 'Обучаем модель на имеющихся данных')
(20, 'Выполняем прогноз временного ряда')
(21, 'get_error_for_model: вход в функцию')
(22, 'get_error_for_model: выход из функци

In [4]:
# выгрузим данные из таблицы events в датафрейм
query = ''' SELECT * 
            FROM events 
        '''
events_data12 = pd.io.sql.read_sql(query, con = con)
events_data12

id                     forecastMarker                request_time  \
0          1           SelfService/RUS/turnover  2022-09-12 11:30:11.106628   
1          2           SelfService/RUS/turnover  2022-09-12 11:30:11.111359   
2          3           SelfService/RUS/turnover  2022-09-12 11:30:11.115989   
3          4           SelfService/RUS/turnover  2022-09-12 11:30:12.761754   
4          5           SelfService/RUS/turnover  2022-09-12 11:30:12.770134   
...      ...                                ...                         ...   
11202  11203  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.360209   
11203  11204  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.367637   
11204  11205  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.377117   
11205  11206  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.382090   
11206  11207  weekly_report.SS:no_segment_users  2022-09-12 18:24:18.389464   

       event_id            sender  \
0             1       test_system   
1            29       test_system   
2             2       test_system   
3             3       test_system   
4             5       test_system   
...         ...               ...   
11202        24  weekly_report.ss   
11203        15  weekly_report.ss   
11204        16  weekly_report.ss   
11205        17  weekly_report.ss   
11206        31  weekly_report.ss   

                                                 context  \
0      df = <class 'NoneType'>, sql_query = <class 'N...   
1                                                   None   
2                                                   None   
3                          Найдено 1 строк с параметрами   
4                                                   None   
...                                                  ...   
11202                                               None   
11203                                               None   
11204                                               None   
11205                                               None   
11206                          Отдан датафрейм: (894, 4)   

                             session_id  
0      6f4277f6df57c4f4e90779966522c876  
1      6f4277f6df57c4f4e90779966522c876  
2      6f4277f6df57c4f4e90779966522c876  
3      6f4277f6df57c4f4e90779966522c876  
4      6f4277f6df57c4f4e90779966522c876  
...                                 ...  
11202  0d2086d69468e0587d193654a3a0b746  
11203  0d2086d69468e0587d193654a3a0b746  
11204  0d2086d69468e0587d193654a3a0b746  
11205  0d2086d69468e0587d193654a3a0b746  
11206  0d2086d69468e0587d193654a3a0b746  

[11207 rows x 7 columns]

In [5]:
# создадим подключение к БД с логами за 13 сентября и выгрузим их в датафрейм
con13 = sl.connect('logs_2022-09-13.db')
query = ''' SELECT * 
            FROM events 
        '''
events_data13 = pd.io.sql.read_sql(query, con = con13)
events_data13

id                     forecastMarker                request_time  \
0          1                               None  2022-09-13 00:10:27.967745   
1          2                               None  2022-09-13 00:10:27.972425   
2          3                               None  2022-09-13 00:10:27.977883   
3          4                               None  2022-09-13 00:10:27.982433   
4          5                               None  2022-09-13 00:10:27.989409   
...      ...                                ...                         ...   
22270  22271  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.052806   
22271  22272  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.058017   
22272  22273  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.065892   
22273  22274  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.071837   
22274  22275  weekly_report.SS:no_segment_users  2022-09-13 09:41:15.078643   

       event_id                 sender  \
0             1                   None   
1            29                   None   
2             5                   None   
3             8                   None   
4             9                   None   
...         ...                    ...   
22270        24  test.weekly_report.ss   
22271        15  test.weekly_report.ss   
22272        16  test.weekly_report.ss   
22273        17  test.weekly_report.ss   
22274        31  test.weekly_report.ss   

                                                 context  \
0      df = <class 'pandas.core.frame.DataFrame'>, sq...   
1                                                   None   
2                                                   None   
3                                                   None   
4                                                   None   
...                                                  ...   
22270                                               None   
22271                                               None   
22272                                               None   
22273                                               None   
22274                          Отдан датафрейм: (895, 4)   

                             session_id  
0      fa89a4d46b9b34761a580f0152f94f05  
1      fa89a4d46b9b34761a580f0152f94f05  
2      fa89a4d46b9b34761a580f0152f94f05  
3      fa89a4d46b9b34761a580f0152f94f05  
4      fa89a4d46b9b34761a580f0152f94f05  
...                                 ...  
22270  d637d1b6d2b70fd088e2ea44959483cb  
22271  d637d1b6d2b70fd088e2ea44959483cb  
22272  d637d1b6d2b70fd088e2ea44959483cb  
22273  d637d1b6d2b70fd088e2ea44959483cb  
22274  d637d1b6d2b70fd088e2ea44959483cb  

[22275 rows x 7 columns]

In [6]:
# создадим подключение к БД с логами за 14 сентября и выгрузим их в датафрейм
con14 = sl.connect('logs_2022-09-14.db')
query = ''' SELECT * 
            FROM events 
        '''
events_data14 = pd.io.sql.read_sql(query, con = con14)
events_data14

id                         forecastMarker  \
0          1                                   None   
1          2                                   None   
2          3                                   None   
3          4                                   None   
4          5                                   None   
...      ...                                    ...   
11244  11245  weekly_report.SS:new_users_90d:tiktok   
11245  11246  weekly_report.SS:new_users_90d:tiktok   
11246  11247  weekly_report.SS:new_users_90d:tiktok   
11247  11248  weekly_report.SS:new_users_90d:tiktok   
11248  11249  weekly_report.SS:new_users_90d:tiktok   

                     request_time  event_id            sender  \
0      2022-09-14 00:09:39.234361         1              None   
1      2022-09-14 00:09:39.238851        29              None   
2      2022-09-14 00:09:39.242626         5              None   
3      2022-09-14 00:09:39.248606         8              None   
4      2022-09-14 00:09:39.254990         9              None   
...                           ...       ...               ...   
11244  2022-09-14 06:07:47.582138        24  weekly_report.ss   
11245  2022-09-14 06:07:47.588185        15  weekly_report.ss   
11246  2022-09-14 06:07:47.594110        16  weekly_report.ss   
11247  2022-09-14 06:07:47.601623        17  weekly_report.ss   
11248  2022-09-14 06:07:47.609898        31  weekly_report.ss   

                                                 context  \
0      df = <class 'pandas.core.frame.DataFrame'>, sq...   
1                                                   None   
2                                                   None   
3                                                   None   
4                                                   None   
...                                                  ...   
11244                                               None   
11245                                               None   
11246                                               None   
11247                                               None   
11248                         Отдан датафрейм: (1352, 4)   

                             session_id  
0      ec6a82e4a1b9b4dc467639eb1a5d6987  
1      ec6a82e4a1b9b4dc467639eb1a5d6987  
2      ec6a82e4a1b9b4dc467639eb1a5d6987  
3      ec6a82e4a1b9b4dc467639eb1a5d6987  
4      ec6a82e4a1b9b4dc467639eb1a5d6987  
...                                 ...  
11244  7caa7e3d9378310a86ce7e20fc58b561  
11245  7caa7e3d9378310a86ce7e20fc58b561  
11246  7caa7e3d9378310a86ce7e20fc58b561  
11247  7caa7e3d9378310a86ce7e20fc58b561  
11248  7caa7e3d9378310a86ce7e20fc58b561  

[11249 rows x 7 columns]

In [7]:
# объединим данные за 3 дня в один датафрейм
events_data = pd.concat([events_data12, events_data13, events_data14])

# Выгрузим датафрейм в CSV-файл
events_data.to_csv('events_data.csv', sep='\t', encoding='utf-8', index=False)

ПОдключение в БД выполнено, логи за три дня объединены в один файл. Количество логов за 13 сентября в 2 раза превышает количество логов в другие дни (12 и 14).

## Обзор данных

In [8]:
events_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44731 entries, 0 to 11248
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              44731 non-null  int64 
 1   forecastMarker  44731 non-null  object
 2   request_time    44731 non-null  object
 3   event_id        44731 non-null  int64 
 4   sender          44731 non-null  object
 5   context         44731 non-null  object
 6   session_id      44731 non-null  object
dtypes: int64(2), object(5)
memory usage: 2.7+ MB


Пропусков в данных нет. Время события хранится не в оптимальном формате. Приветем к типу datetime

In [9]:
events_data['request_time'] = pd.to_datetime(events_data['request_time'], format='%Y-%m-%d %H:%M:%S.%f')

In [10]:
events_data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 44731 entries, 0 to 11248
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id              44731 non-null  int64         
 1   forecastMarker  44731 non-null  object        
 2   request_time    44731 non-null  datetime64[ns]
 3   event_id        44731 non-null  int64         
 4   sender          44731 non-null  object        
 5   context         44731 non-null  object        
 6   session_id      44731 non-null  object        
dtypes: datetime64[ns](1), int64(2), object(4)
memory usage: 2.7+ MB


In [11]:
for col in events_data.columns:
    print(col)
    print(events_data[col].nunique())
    print(events_data[col].sort_values().unique())

id
22275
[    1     2     3 ... 22273 22274 22275]
forecastMarker
366
['Agency/NewClients' 'Agency/turnover' 'None' 'SaaS/RUS/activation'
 'SaaS/RUS/turnover' 'SaaS/RUS/turnover/avito'
 'SaaS/RUS/turnover/direct_internal' 'SaaS/RUS/turnover/google_internal'
 'SaaS/RUS/turnover/targetMailRu' 'SaaS/RUS/turnover/vkontakte'
 'SelfService/RUS/activ' 'SelfService/RUS/activation'
 'SelfService/RUS/ag_core' 'SelfService/RUS/ag_new'
 'SelfService/RUS/ag_turnover' 'SelfService/RUS/core'
 'SelfService/RUS/reg' 'SelfService/RUS/turnover' 'SubAgency/RUS/activ'
 'SubAgency/RUS/activation' 'SubAgency/RUS/ag_core' 'SubAgency/RUS/ag_new'
 'SubAgency/RUS/ag_turnover' 'SubAgency/RUS/core' 'SubAgency/RUS/reg'
 'SubAgency/RUS/turnover' 'Total/Turnover' 'a.shakhtarina@elama.ru'
 'all@users' 'e.kalugin@elama.ru' 'n.ezhova@elama.ru'
 'u.malyshenko@elama.ru' 'weekly_report.SA:activations_clients'
 'weekly_report.SA:activations_groups'
 'weekly_report.SA:new_agencies_90d:all_systems'
 'weekly_report.SA:new_agen

1599
['0035de8fad557cfe0b187f24b954195f' '004df69bde1839cb768d2003ff28ec6c'
 '0090da9b342c3e895cd77bfa1cd747f9' ... 'ff7de558c7e51213dcf10998411618fb'
 'ff8ae4d9f46d819134f579453e2bb786' 'ffab4186fa0ad7a045cb7daa19883410']


За три исследуемых дня:
- id дублируются, как как логи за каждый день имеют свою собственную нумерацию событий, начинающуюся с 1.
- было поставлено 365 задач прогнозирования конкретных бизнесс-показателей (часть задач не имеет forecastMarker - значение 'None')  от 34 систем, поставивших задачу (часть задач не имеют отметку sender - значение 'None').
- всего обработано 1599 задач прогнозирования.
- время начала каждого события уникально, одновременно два события не начинались.
- за три дня ни разу не произошло событие с идентификатором 30 - Пустой набор данных для прогнозирования и 33 - get_error_for_model: прошли cross_validation.

In [12]:
events_data.groupby(by='forecastMarker')['id'].count().sort_values(ascending=False)

forecastMarker
None                                                          4139
weekly_report.SAAS:new_visitors:utm_medium:cpc_branded         393
weekly_report.SAAS:new_visitors:utm_medium:messengers          249
weekly_report.SAAS:new_visitors:utm_medium:cpc_not_branded     247
weekly_report.SAAS:new_visitors:utm_medium:email               247
                                                              ... 
SubAgency/RUS/ag_core                                           52
SubAgency/RUS/ag_new                                            52
SubAgency/RUS/core                                              52
SubAgency/RUS/reg                                               52
SubAgency/RUS/ag_turnover                                       52
Name: id, Length: 366, dtype: int64

Наиболее часто обрабатываюся задачи без forecastMarker. Далее следуют "weekly_report.SAAS:new_visitors:utm_medium:cpc_branded" и "weekly_report.SAAS:new_visitors:utm_medium:messengers".

Каждая задача обрабатывалась за три дня минимум 52 раза.

In [13]:
events_data.groupby(by='sender')['id'].count().sort_values(ascending=False)

sender
weekly_report.care                                                                          10039
weekly_report.sa                                                                             8809
weekly_report.ss                                                                             8732
bq_analytics_tables_refresh                                                                  3452
test.weekly_report.care                                                                      2986
test.weekly_report.sa                                                                        2522
test.weekly_report.ss                                                                        2496
weekly_report.saas                                                                           1622
weekly_report.saas_monthly                                                                    831
forecast_for_PHR.load_prophet_forecast_data                                                   732
DAG : bq_anal

Часть задач пронозирования имеют тестовые версии.

Наиболее часто задачу ставят следующие системы 
- weekly_report.care (10 000 раз за исследуемый период),
- weekly_report.sa (8 800 раз за исследуемый период),
- weekly_report.ss (8 700 раз за исследуемый период).

Рекурсивные задачи прогнозирования internal_call - самые редкие, не более 30 раз каждая.

In [14]:
events_data.groupby(by='event_id')['id'].count().sort_values(ascending=False)

event_id
1     1599
8     1599
25    1599
9     1599
10    1599
5     1599
29    1599
27    1598
26    1598
20    1598
11    1598
12    1598
13    1598
14    1598
18    1598
19    1598
31    1594
24    1594
23    1594
17    1594
16    1594
15    1594
7     1549
6     1549
2     1442
3     1414
22    1163
21     812
32     812
34     809
4       28
28      14
Name: id, dtype: int64

За исследуемый период 1599 задач началось и 1594 задачи закончилось. 5 задач не завершилось успешно.

Наиболее редко при решении задачи прогнозирования происходят события 28 "getForecastMarkerAsRegressor: выход из функции" и 4 "Выход из функции getParametrsAndRegressors БЕЗ параметров". Вероятно поэтому более 99.6% задач завершились успешно.

События 5, 8, 9, 10, 25, 29 наступали при решении каждой задачи прогнозирования. А события 21, 32 и 34 только в 50% задач.

In [15]:
# добавим объединенный датафрейм с логами за три дня в БД
events_data.to_sql('events_all', con, if_exists='append')

44731

##  Вопросы тестового задания.

In [16]:
con12 = sl.connect('logs_2022-09-12.db')

### Между какими событиями наибольший шаг? Укажите пару идентификаторов событий с наибольшим шагом, относительно всей базы (всех файлов).


In [17]:
query = ''' WITH tab as
            (SELECT event_id,
                   session_id,
                   request_time,
                   LEAD(event_id) OVER(PARTITION BY session_id ORDER BY request_time) as next_event,
                   LEAD(request_time) OVER(PARTITION BY session_id ORDER BY request_time) as next_time
              FROM events_all)
              SELECT *,
              ROUND((JULIANDAY(next_time) - JULIANDAY(request_time)) * 86400, 5) as event_duration
              FROM tab
              ORDER BY event_duration DESC            
        '''
event_data1 = pd.io.sql.read_sql(query, con = con12)
event_data1

event_id                        session_id  \
0              6  4dcab2af4573ced8251fb7433174ba15   
1              6  7714ae1e594b3b4cb64cae7a4607bcce   
2              6  e35e3228f80531fbc06572a0baf07c11   
3              6  f4cd592c2ad8eb3db4d13376c46f9a49   
4              6  e55a726e9ec99edc7448d2fd25867d83   
...          ...                               ...   
447305        31  ff46d80e496dcf619510c8afe9a4696e   
447306        31  ff73ba77cc519c9731001586b3cf56ad   
447307        31  ff7de558c7e51213dcf10998411618fb   
447308        31  ff8ae4d9f46d819134f579453e2bb786   
447309        31  ffab4186fa0ad7a045cb7daa19883410   

                      request_time  next_event                   next_time  \
0       2022-09-14 05:20:53.759918         7.0  2022-09-14 05:22:40.308128   
1       2022-09-14 05:18:50.881412         7.0  2022-09-14 05:19:55.017136   
2       2022-09-13 08:59:34.661514         7.0  2022-09-13 09:00:38.689271   
3       2022-09-12 17:14:50.071812         7.0  2022-09-12 17:15:40.349360   
4       2022-09-12 17:36:03.328557         7.0  2022-09-12 17:36:52.877005   
...                            ...         ...                         ...   
447305  2022-09-14 05:42:59.804307         NaN                        None   
447306  2022-09-12 17:33:10.967604         NaN                        None   
447307  2022-09-14 05:13:51.300524         NaN                        None   
447308  2022-09-14 05:30:53.151991         NaN                        None   
447309  2022-09-13 08:44:24.810658         NaN                        None   

        event_duration  
0            106.54897  
1             64.13600  
2             64.02798  
3             50.27802  
4             49.54900  
...                ...  
447305             NaN  
447306             NaN  
447307             NaN  
447308             NaN  
447309             NaN  

[447310 rows x 6 columns]

Наибольший шаг за три дня зафиксирован между событиями 6 "Начинаем читать данные из BQ по полученному SQL запросу" и 7 "Получили данные из BQ по полученному SQL запросу" в сессии 4dcab2af4573ced8251fb7433174ba15. Он составил более 106 секунд.            

### Интервал времени до наступления какого события показывает наибольший разброс. Укажите пару идентификатор и sender с наибольшим разбросом. Предложите свои идей, с чем это может быть связано.


In [18]:
query = ''' WITH tab as
            (SELECT event_id,
                   session_id,
                   sender,
                   request_time,
                   LEAD(event_id) OVER(PARTITION BY session_id ORDER BY request_time) as next_event,
                   LEAD(request_time) OVER(PARTITION BY session_id ORDER BY request_time) as next_time
              FROM events_all),
              tab1 as
              (SELECT *,
              ROUND((JULIANDAY(next_time) - JULIANDAY(request_time)) * 86400, 5) as event_duration
              FROM tab
              ORDER BY event_duration DESC)
             
             SELECT 
                    sender,
                    next_event,
                    MAX(event_duration) - MIN(event_duration) as spread_of_values
             FROM tab1
             WHERE event_duration > 0
             GROUP BY next_event, sender
             ORDER BY spread_of_values DESC
        '''
event_data2 = pd.io.sql.read_sql(query, con = con12)
event_data2

sender  next_event  \
0                                   weekly_report.care           7   
1                              test.weekly_report.care           7   
2                                   weekly_report.saas           7   
3                                     weekly_report.sa           7   
4                                test.weekly_report.sa           7   
..                                                 ...         ...   
941  internal_call for SelfService/RUS/turnover, se...          34   
942  internal_call for SelfService/RUS/turnover, se...          34   
943  internal_call for SubAgency/RUS/turnover, sess...          34   
944  internal_call for SubAgency/RUS/turnover, sess...          34   
945                                        test_system          34   

     spread_of_values  
0           105.00897  
1            62.55999  
2            37.04498  
3            36.74098  
4            36.58399  
..                ...  
941           0.00000  
942           0.00000  
943           0.00000  
944           0.00000  
945           0.00000  

[946 rows x 3 columns]

Наибольший разброс в значениях имеет интервал времени до наступления события 7 "Получили данные из BQ по полученному SQL запросу" при постановке задачи weekly_report.care. Он составляет более 105 секунд. Вероятно на это влияет размер считываемых данных: чем больший объем считываем, тем дольше выполняется операция.

### Какая задача прогнозирования выполняется дольше всего (название задачи)?

In [19]:
query = ''' SELECT forecastMarker, 
                   session_id,
                   strftime('%s',  max) - strftime('%s', min) as event_duration
            FROM (SELECT forecastMarker, 
                         session_id,
                         max(request_time) as max,
                         min(request_time) as min
                    FROM events_all
                GROUP BY session_id)
            ORDER BY event_duration DESC
        '''
event_data3 = pd.io.sql.read_sql(query, con = con12)
event_data3


forecastMarker                        session_id  \
0     weekly_report.care:count_of_jivo_SA  4dcab2af4573ced8251fb7433174ba15   
1             weekly_report.SAAS:visitors  f151aa3db81db96c276313c910abca75   
2        weekly_report.care:sum_of_income  7714ae1e594b3b4cb64cae7a4607bcce   
3             weekly_report.SAAS:visitors  f96f949df754e973ae7ed08f79c64c6c   
4     weekly_report.care:count_of_jivo_SS  e35e3228f80531fbc06572a0baf07c11   
...                                   ...                               ...   
1594                                 None  3ce3ef68895958bddf308c4a86d75f69   
1595                                 None  cb8700fb5bf851b3f3bfcccefafc36e5   
1596                                 None  9e643e715f004a7c273a41c71c2d3777   
1597                                 None  b749fd567a0e5aa231247cd83ea7aee6   
1598                                 None  ec6a82e4a1b9b4dc467639eb1a5d6987   

      event_duration  
0                112  
1                 73  
2                 72  
3                 71  
4                 70  
...              ...  
1594               5  
1595               5  
1596               4  
1597               4  
1598               4  

[1599 rows x 3 columns]

In [20]:
query = ''' SELECT *
              FROM events_all
            WHERE forecastMarker = 'weekly_report.care:count_of_jivo_SA'
                  AND session_id = '4dcab2af4573ced8251fb7433174ba15'
            ORDER BY request_time DESC
        '''
event_data = pd.io.sql.read_sql(query, con = con12)
event_data

index    id                       forecastMarker  \
0     5762  5763  weekly_report.care:count_of_jivo_SA   
1     5762  5763  weekly_report.care:count_of_jivo_SA   
2     5762  5763  weekly_report.care:count_of_jivo_SA   
3     5762  5763  weekly_report.care:count_of_jivo_SA   
4     5762  5763  weekly_report.care:count_of_jivo_SA   
..     ...   ...                                  ...   
255   5737  5738  weekly_report.care:count_of_jivo_SA   
256   5737  5738  weekly_report.care:count_of_jivo_SA   
257   5737  5738  weekly_report.care:count_of_jivo_SA   
258   5737  5738  weekly_report.care:count_of_jivo_SA   
259   5737  5738  weekly_report.care:count_of_jivo_SA   

                   request_time  event_id              sender  \
0    2022-09-14 05:22:44.688175        31  weekly_report.care   
1    2022-09-14 05:22:44.688175        31  weekly_report.care   
2    2022-09-14 05:22:44.688175        31  weekly_report.care   
3    2022-09-14 05:22:44.688175        31  weekly_report.care   
4    2022-09-14 05:22:44.688175        31  weekly_report.care   
..                          ...       ...                 ...   
255  2022-09-14 05:20:52.203702         1  weekly_report.care   
256  2022-09-14 05:20:52.203702         1  weekly_report.care   
257  2022-09-14 05:20:52.203702         1  weekly_report.care   
258  2022-09-14 05:20:52.203702         1  weekly_report.care   
259  2022-09-14 05:20:52.203702         1  weekly_report.care   

                                               context  \
0                            Отдан датафрейм: (913, 4)   
1                            Отдан датафрейм: (913, 4)   
2                            Отдан датафрейм: (913, 4)   
3                            Отдан датафрейм: (913, 4)   
4                            Отдан датафрейм: (913, 4)   
..                                                 ...   
255  df = <class 'NoneType'>, sql_query = <class 'N...   
256  df = <class 'NoneType'>, sql_query = <class 'N...   
257  df = <class 'NoneType'>, sql_query = <class 'N...   
258  df = <class 'NoneType'>, sql_query = <class 'N...   
259  df = <class 'NoneType'>, sql_query = <class 'N...   

                           session_id  
0    4dcab2af4573ced8251fb7433174ba15  
1    4dcab2af4573ced8251fb7433174ba15  
2    4dcab2af4573ced8251fb7433174ba15  
3    4dcab2af4573ced8251fb7433174ba15  
4    4dcab2af4573ced8251fb7433174ba15  
..                                ...  
255  4dcab2af4573ced8251fb7433174ba15  
256  4dcab2af4573ced8251fb7433174ba15  
257  4dcab2af4573ced8251fb7433174ba15  
258  4dcab2af4573ced8251fb7433174ba15  
259  4dcab2af4573ced8251fb7433174ba15  

[260 rows x 8 columns]

Задача прогнозирования weekly_report.care:count_of_jivo_SA выполняется дольше всех - 112 секунд.

### Какое количество задач прогнозирования могут выполняться одновременно? Укажите максимальное число параллельных задач.

In [21]:
query = ''' SELECT session_id,
                   min(request_time) as session_start,
                   max(request_time) as session_end
              FROM events_all
          GROUP BY session_id          
        '''
event_data4 = pd.io.sql.read_sql(query, con = con12)
event_data4

session_id               session_start  \
0     0035de8fad557cfe0b187f24b954195f  2022-09-12 17:11:29.821083   
1     004df69bde1839cb768d2003ff28ec6c  2022-09-14 05:39:44.196050   
2     0090da9b342c3e895cd77bfa1cd747f9  2022-09-14 05:44:10.382877   
3     00d3543ff8b2d409f3ec4e26cdc246d2  2022-09-14 06:03:25.665344   
4     00fd7becb50ab7fddd284b633a68be4f  2022-09-12 18:09:22.917658   
...                                ...                         ...   
1594  ff46d80e496dcf619510c8afe9a4696e  2022-09-14 05:42:40.334153   
1595  ff73ba77cc519c9731001586b3cf56ad  2022-09-12 17:33:00.784616   
1596  ff7de558c7e51213dcf10998411618fb  2022-09-14 05:13:39.279998   
1597  ff8ae4d9f46d819134f579453e2bb786  2022-09-14 05:30:43.654128   
1598  ffab4186fa0ad7a045cb7daa19883410  2022-09-13 08:44:16.092260   

                     session_end  
0     2022-09-12 17:11:39.209815  
1     2022-09-14 05:39:54.379904  
2     2022-09-14 05:44:40.509244  
3     2022-09-14 06:03:52.605758  
4     2022-09-12 18:09:34.547063  
...                          ...  
1594  2022-09-14 05:42:59.804307  
1595  2022-09-12 17:33:10.967604  
1596  2022-09-14 05:13:51.300524  
1597  2022-09-14 05:30:53.151991  
1598  2022-09-13 08:44:24.810658  

[1599 rows x 3 columns]

In [22]:
# создадим массив времен и отсортируем его
times = pd.concat([event_data4['session_start'], event_data4['session_end']])\
        .sort_values().reset_index(drop=True).rename('time')

# переведем в формат хранения времени
times = pd.to_datetime(times)

# создаем вспомогательный столбец следующих времен, обновляем индекс и объединяем в датафрейм
times1 = pd.Series(times[1:], name='next_time').reset_index(drop=True)
times_df = pd.concat([times, times1], axis=1)

# расчитываем среднее между двумя точками времени
times_df['delta'] = times_df['time'] + (times_df['next_time'] - times_df['time'])/2
times_df

time                  next_time  \
0    2022-09-12 11:30:11.106628 2022-09-12 11:30:18.666983   
1    2022-09-12 11:30:18.666983 2022-09-12 11:30:29.936695   
2    2022-09-12 11:30:29.936695 2022-09-12 11:30:36.249818   
3    2022-09-12 11:30:36.249818 2022-09-12 12:13:46.808490   
4    2022-09-12 12:13:46.808490 2022-09-12 12:13:57.512095   
...                         ...                        ...   
3193 2022-09-14 06:07:26.420862 2022-09-14 06:07:26.921552   
3194 2022-09-14 06:07:26.921552 2022-09-14 06:07:36.054822   
3195 2022-09-14 06:07:36.054822 2022-09-14 06:07:36.558670   
3196 2022-09-14 06:07:36.558670 2022-09-14 06:07:47.609898   
3197 2022-09-14 06:07:47.609898                        NaT   

                             delta  
0    2022-09-12 11:30:14.886805500  
1    2022-09-12 11:30:24.301839000  
2    2022-09-12 11:30:33.093256500  
3    2022-09-12 11:52:11.529154000  
4    2022-09-12 12:13:52.160292500  
...                            ...  
3193 2022-09-14 06:07:26.671207000  
3194 2022-09-14 06:07:31.488187000  
3195 2022-09-14 06:07:36.306746000  
3196 2022-09-14 06:07:42.084284000  
3197                           NaT  

[3198 rows x 3 columns]

In [23]:
times_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3198 entries, 0 to 3197
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   time       3198 non-null   datetime64[ns]
 1   next_time  3197 non-null   datetime64[ns]
 2   delta      3197 non-null   datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 99.9 KB


In [24]:
event_data4['session_end'] = pd.to_datetime(event_data4['session_end'])
event_data4['session_start'] = pd.to_datetime(event_data4['session_start'])
event_data4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   session_id     1599 non-null   object        
 1   session_start  1599 non-null   datetime64[ns]
 2   session_end    1599 non-null   datetime64[ns]
dtypes: datetime64[ns](2), object(1)
memory usage: 37.6+ KB


In [25]:
session_count = []
for i in tqdm(range(times_df.shape[0])):
    sc = 0
    for j in range(event_data4.shape[0]):
        if (event_data4['session_start'][j] <= times_df['delta'][i]) & (event_data4['session_end'][j] >= times_df['delta'][i]):
            sc +=1
    session_count.append(sc)
    #session_count[i] = sc
max(session_count)

100%|███████████████████████████████████████| 3198/3198 [05:45<00:00,  9.25it/s]


6

Алгоритм отрабатывает очень долго. Попробуем применить библиотеку pandas.

In [26]:
session_count = []
for i in tqdm(range(times_df.shape[0])):
    ev_df1=event_data4[event_data4['session_start'] <= times_df['delta'][i]]
    ev_df2=ev_df1[ev_df1['session_end'] >= times_df['delta'][i]]
    sc = ev_df2.shape[0]
    session_count.append(sc)
    #session_count[i] = sc
max(session_count)

100%|█████████████████████████████████████| 3198/3198 [00:03<00:00, 1006.57it/s]


6

Максимальное количество задач, обрабатываемых одновременно - 6.

Альтернативное решение (проверим себя). Создадим датафрейм со всеми временными метками и флагом начало\конец выполнения задачи.

In [27]:
d1={'time': event_data4['session_start']}
df1 = pd.DataFrame(data=d1)
df1['start'] = 1
df1.head()

time  start
0 2022-09-12 17:11:29.821083      1
1 2022-09-14 05:39:44.196050      1
2 2022-09-14 05:44:10.382877      1
3 2022-09-14 06:03:25.665344      1
4 2022-09-12 18:09:22.917658      1

In [28]:
d2={'time': event_data4['session_end']}
df2 = pd.DataFrame(data=d2)
df2['start'] = 0
df2.head()

time  start
0 2022-09-12 17:11:39.209815      0
1 2022-09-14 05:39:54.379904      0
2 2022-09-14 05:44:40.509244      0
3 2022-09-14 06:03:52.605758      0
4 2022-09-12 18:09:34.547063      0

In [29]:
df = pd.concat([df1, df2], ignore_index=True)
df = df.sort_values(by='time').reset_index(drop=True)
df.head()

time  start
0 2022-09-12 11:30:11.106628      1
1 2022-09-12 11:30:18.666983      1
2 2022-09-12 11:30:29.936695      0
3 2022-09-12 11:30:36.249818      0
4 2022-09-12 12:13:46.808490      1

Если задача начинается прибавим 1, если задача закончена вычтем 1. 

In [30]:
%%time
task_count = 0
max_count = 0
for i in range(df.shape[0]):
    if df['start'][i] == 1:
        task_count +=1
    else:
        task_count -=1
    if task_count > max_count:
        max_count = task_count
max_count

CPU times: user 22.1 ms, sys: 1.2 ms, total: 23.3 ms
Wall time: 23.4 ms


6

### Перечислите forecastMarker которые не завершились успешно.

In [31]:
query = ''' WITH tab as
            (SELECT session_id
              FROM events_all
             WHERE event_id = 31)
            SELECT forecastMarker,
                   session_id
            FROM events_all
            WHERE session_id NOT IN tab
            GROUP BY forecastMarker, session_id
        '''
event_data5 = pd.io.sql.read_sql(query, con = con12)
event_data5

forecastMarker  \
0                                               None   
1                     SaaS/RUS/turnover/targetMailRu   
2                    weekly_report.SAAS:new_visitors   
3  weekly_report.SAAS:new_visitors:utm_medium:org...   
4                        weekly_report.SAAS:turnover   

                         session_id  
0  56af71228bfbb91afd00a04652646069  
1  05fde45de80ea82b620d924acf35410c  
2  ef2ce869aedbddee01530295d44a5993  
3  56e957cb0972c856c211dfed092dc8c2  
4  9dd140e965fe9c98656a0d0e405aa1e8

За 12-14 сентября 5 задач прогнозирования не были завершены успешно (событие с индексом 31 не наступило). Это задачи с forecastMarker "SaaS/RUS/turnover/targetMailRu", "weekly_report.SAAS:new_visitors", "weekly_report.SAAS:new_visitors:utm_medium:org...", "weekly_report.SAAS:turnover" и одна задача с session_id "56af71228bfbb91afd00a04652646069".

## Итоги исследования.


Мы получили от заказчика три БД, в которых хранятся логи событий задач прогнозирования за 12, 13 и 14 сентября. Объединили таблицы с логами в один датафрейм и обновили БД.

Количество логов за 13 сентября в 2 раза превышает количество логов в другие дни (12 и 14). При этом данные у нас за понедельник, вторник и среду. Вероятно это выраженная недельная сезонность, а возможно и сбой в логировании событий.

За три исследуемых дня:
- id дублируются, как как логи за каждый день имеют свою собственную нумерацию событий, начинающуюся с 1.
- было поставлено 365 задач прогнозирования конкретных бизнесс-показателей (часть задач не имеет forecastMarker - значение 'None')  от 34 систем, поставивших задачу (часть задач не имеют отметку sender - значение 'None').
- всего обработано 1599 задач прогнозирования (session_id).
- время начала каждого события уникально, одновременно два события не начинались.
- за три дня ни разу не произошло событие с идентификатором 30 - Пустой набор данных для прогнозирования и 33 - get_error_for_model: прошли cross_validation.

Наиболее часто обрабатываюся задачи без forecastMarker. Далее следуют "weekly_report.SAAS:new_visitors:utm_medium:cpc_branded" и "weekly_report.SAAS:new_visitors:utm_medium:messengers".

Каждая задача конкретного бизнесс-показателя обрабатывалась за три дня минимум 52 раза, максимум 393 (в зависимости от forecastMarker).

Часть задач пронозирования имеют тестовые версии.

Наиболее часто задачу ставят следующие системы (sender )
- weekly_report.care (10 000 раз за исследуемый период),
- weekly_report.sa (8 800 раз за исследуемый период),
- weekly_report.ss (8 700 раз за исследуемый период).

Рекурсивные задачи прогнозирования internal_call - самые редкие, не более 30 раз каждая.

За исследуемый период 1599 задач началось и 1594 задачи закончилось. 5 задач не завершилось успешно.

Наиболее редко при решении задачи прогнозирования происходят события 28 "getForecastMarkerAsRegressor: выход из функции" и 4 "Выход из функции getParametrsAndRegressors БЕЗ параметров". Вероятно поэтому более 99.6% задач завершились успешно.

События 5, 8, 9, 10, 25, 29 наступали при решении каждой задачи прогнозирования. А события 21, 32 и 34 только в 50% задач.

**1. Между какими событиями наибольший шаг? Укажите пару идентификаторов событий с наибольшим шагом, относительно всей базы (всех файлов).**
- Наибольший шаг за три дня зафиксирован между событиями 6 "Начинаем читать данные из BQ по полученному SQL запросу" и 7 "Получили данные из BQ по полученному SQL запросу" в сессии 4dcab2af4573ced8251fb7433174ba15. Он составил более 106 секунд.

**2. Интервал времени до наступления какого события показывает наибольший разброс. Укажите пару идентификатор и sender с наибольшим разбросом. Предложите свои идей, с чем это может быть связано.**
- Наибольший разброс в значениях имеет интервал времени до наступления события 7 "Получили данные из BQ по полученному SQL запросу" при постановке задачи weekly_report.care. Он составляет более 105 секунд. Вероятно на это влияет размер считываемых данных: чем больший объем считываем, тем дольше выполняется операция.

**3. Какая задача прогнозирования выполняется дольше всего (название задачи)?**
- Задача прогнозирования weekly_report.care:count_of_jivo_SA выполняется дольше всех - 112 секунд.

**4. Какое количество задач прогнозирования могут выполняться одновременно? Укажите максимальное число параллельных задач.**
- в течение исследуемых трех дней одновременно максимум выполнялось 6 задач прогнозирования.

**5. Перечислите forecastMarker которые не завершились успешно.**
- За 12-14 сентября 5 задач прогнозирования не были завершены успешно (событие с индексом 31 не наступило). 

Это  4 задачи с forecastMarker 
- "SaaS/RUS/turnover/targetMailRu", 
- "weekly_report.SAAS:new_visitors", 
- "weekly_report.SAAS:new_visitors:utm_medium:org...", 
- "weekly_report.SAAS:turnover" .

и 1 задача с session_id  
- "56af71228bfbb91afd00a04652646069".
